File: applications/rtkwaveletsdenoising/rtkwaveletsdenoising.cxx
/*=========================================================================
 *
 *  Copyright RTK Consortium
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *         https://www.apache.org/licenses/LICENSE-2.0.txt
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *=========================================================================*/

#include "rtkwaveletsdenoising_ggo.h"
#include "rtkGgoFunctions.h"

#include "rtkDeconstructSoftThresholdReconstructImageFilter.h"
#include <itkImageFileWriter.h>
#include <itkImageFileReader.h>
#include <string>
#include <sstream>

int
main(int argc, char * argv[])
{
  GGO(rtkwaveletsdenoising, args_info);

  using OutputPixelType = float;
  constexpr unsigned int Dimension = 3;

  using OutputImageType = itk::Image<OutputPixelType, Dimension>;

  // Read the input image
  OutputImageType::Pointer input;
  TRY_AND_EXIT_ON_ITK_EXCEPTION(input = itk::ReadImage<OutputImageType>(args_info.input_arg))

  // Create the denoising filter
  using WaveletsSoftThresholdFilterType = rtk::DeconstructSoftThresholdReconstructImageFilter<OutputImageType>;
  auto wst = WaveletsSoftThresholdFilterType::New();
  wst->SetInput(input);
  wst->SetOrder(args_info.order_arg);
  wst->SetThreshold(args_info.threshold_arg);
  wst->SetNumberOfLevels(args_info.level_arg);

  // Write reconstruction
  TRY_AND_EXIT_ON_ITK_EXCEPTION(itk::WriteImage(wst->GetOutput(), args_info.output_arg))

  return EXIT_SUCCESS;
}

File: applications/rtkwaveletsdenoising/rtkwaveletsdenoising.py
#!/usr/bin/env python
import argparse
import itk
from itk import RTK as rtk


def build_parser():
    parser = rtk.RTKArgumentParser(
        description="Denoises a volume using Daubechies wavelets soft thresholding"
    )

    parser.add_argument(
        "--input", "-i", help="Input file name", type=str, required=True
    )
    parser.add_argument(
        "--output", "-o", help="Output file name", type=str, required=True
    )
    parser.add_argument(
        "--order", help="Order of the Daubechies wavelets", type=int, required=True
    )
    parser.add_argument(
        "--level",
        "-l",
        help="Number of deconstruction levels",
        type=int,
        required=True,
    )
    parser.add_argument(
        "--threshold",
        "-t",
        help="Threshold used in soft thresholding of the wavelets coefficients",
        type=float,
        required=True,
    )

    return parser


def process(args_info: argparse.Namespace):
    OutputPixelType = itk.F
    Dimension = 3
    OutputImageType = itk.Image[OutputPixelType, Dimension]

    # Read the input image
    if args_info.verbose:
        print(f"Reading input from {args_info.input}...")
    # Read as float to match the filter's expected image type
    input_img = itk.imread(args_info.input, OutputPixelType)

    # Create the denoising filter
    wst = rtk.DeconstructSoftThresholdReconstructImageFilter[OutputImageType].New()
    wst.SetInput(input_img)
    wst.SetOrder(args_info.order)
    wst.SetThreshold(args_info.threshold)
    wst.SetNumberOfLevels(args_info.level)

    # Write reconstruction
    if args_info.verbose:
        print(f"Writing output to {args_info.output}...")
    itk.imwrite(wst.GetOutput(), args_info.output)


def main(argv=None):
    parser = build_parser()
    args_info = parser.parse_args(argv)
    process(args_info)


if __name__ == "__main__":
    main()
